File: Depth-First Search/Combinations/String/LET_no44_Wildcard Matching.cpp
class Solution {
private:
    int m, n;
public:
    bool isMatch(string s, string p) {
        m = s.size();
        n = p.size();
        unordered_map<int, bool> memo;
        return helper(s, 0, p, 0, memo);
    }

    bool helper(string& s, int i, string& p, int j, unordered_map<int, bool>& memo) {
        int key = i * n + j;
        if (memo.count(key))
            return memo[key];
        
        if (i == m) {
            if (j == n)
                return true;
            else {
                while (j < n && p[j] == '*')
                    j++;
                return j == n;
            }
        }

        if (j == n)
            return false;

        bool matched;
        if (p[j] != '*')
            matched = (s[i] == p[j] || p[j] == '?') && helper(s, i + 1, p, j + 1, memo);
        else
            matched = helper(s, i, p, j + 1, memo) || helper(s, i + 1, p, j, memo);

        memo[key] = matched;
        return matched;
    }
};

File: Depth-First Search/Combinations/String/LET_no44_Wildcard Matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        return self.helper(s, p, 0, 0, dict())

    def helper(self, s, p, i, j, memo):
        if (i, j) in memo:
            return memo[(i, j)]

        if i == len(s):
            if j == len(p):
                return True
            else:
                return set(p[j:]) == set('*')
        
        if j == len(p):
            return False

        res = False
        if p[j] == '*':
            res = self.helper(s, p, i + 1, j, memo) or self.helper(s, p, i, j + 1, memo)
        else:
            res = (p[j] in ('?', s[i])) and self.helper(s, p, i + 1, j + 1, memo)
        memo[(i, j)] = res

        return res
    